integration/container/rename_test.go

```diff
@@ -0,0 +1,90 @@
+package container
+
+import (
+	"context"
+	"testing"
+
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/api/types/network"
+	"github.com/docker/docker/api/types/strslice"
+	"github.com/docker/docker/client"
+	"github.com/docker/docker/integration/util/request"
+	"github.com/stretchr/testify/assert"
+)
+
+func runContainer(t *testing.T, client client.APIClient, ctx context.Context, cntCfg *container.Config, hstCfg *container.HostConfig, nwkCfg *network.NetworkingConfig, cntName string) string {
+	cnt, err := client.ContainerCreate(ctx, cntCfg, hstCfg, nwkCfg, cntName)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if err := client.ContainerStart(ctx, cnt.ID, types.ContainerStartOptions{}); err != nil {
+		t.Fatal(err)
+	}
+	return cnt.ID
+}
+
+// This test simulates the scenario mentioned in #31392:
+// Having two linked container, renaming the target and bringing a replacement
+// and then deleting and recreating the source container linked to the new target.
+// This checks that "rename" updates source container correctly and doesn't set it to null.
+func TestRenameLinkedContainer(t *testing.T) {
+	defer setupTest(t)()
```

Comment:
We should add a description to `setupTest` that it returns a cleanup function (and possibly an example). I had to look twice here as I thought this had to be some `teardownTest()` instead 😅

Reply:
Ya, I guess `setupTeardownTest()` would be a better name

Reply:
Alternatively, something like this could be more clear:

```go
fnClean := setupTest(t)
defer fnClean()
```

Reply:
but twice as many lines of code!